### PR TITLE
[roosterteeth] Download ad-free streams

### LINKED ISSUE #7647: Roosterteeth.com ad filter
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2023.07.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Example URLs

https://roosterteeth.com/watch/death-battle-season-10-antman

### Provide a description that is worded well enough to be understood

Some Roosterteeth videos, for example Death Battle season 10, embed ads both at the beginning and in the middle, which I'd like to remove.

Analysis of the HLS M3U files show that the ads are groups of segments, separated from the main video (and each other) by `#EXT-X-DISCONTINUITY` tags.

The existing option `--hls-split-discontinuity` doesn't work well here. While the ads are removed, so is the first half of the video.

I've added a filter of sorts in `downloader/hls.py`, enabled by a new option `--roosterteeth-ad-filter`, that removes these ads on the assumption that each is shorter than 10% of the whole video. My fork is here: https://github.com/Masken3/yt-dlp/tree/roosterteeth_ad_filter

The output below shows a successful run with this option.

I've opened this issue because I'm new to yt-dlp development and would like the maintainers' input on the matter before I open a pull request.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [x] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
> python -m yt_dlp --no-check-certificate -vU --roosterteeth-ad-filter https://roosterteeth.com/watch/death-battle-season-10-antman
[debug] Command-line config: ['--no-check-certificate', '-vU', '--roosterteeth-ad-filter', 'https://roosterteeth.com/watch/death-battle-season-10-antman']
[debug] Encodings: locale cp65001, fs utf-8, pref cp65001, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.07.06 [b532a3481] (source)
[debug] Lazy loading extractors is disabled
[debug] Git HEAD: da85e4757
[debug] Python 3.9.6 (CPython AMD64 64bit) - Windows-10-10.0.17763-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 2022-02-14-git-59c647bcf3-essentials_build-www.gyan.dev (setts), ffprobe 2022-02-14-git-59c647bcf3-essentials_build-www.gyan.dev
[debug] Optional libraries: certifi-2021.05.30, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1855 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.07.06, Current version: stable@2023.07.06
yt-dlp is up to date (stable@2023.07.06)
[RoosterTeeth] Extracting URL: https://roosterteeth.com/watch/death-battle-season-10-antman
[RoosterTeeth] death-battle-season-10-antman: Downloading video JSON metadata
[RoosterTeeth] death-battle-season-10-antman: Downloading m3u8 information
[RoosterTeeth] death-battle-season-10-antman: Downloading episode JSON metadata
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 50934: Downloading 1 format(s): hls-4606+hls-audio-3-en__Main_
[debug] Invoking hlsnative downloader on "https://ssaimanifest.prod.boltdns.net/us-east-1/playback/once/v1/hls/v4/clear/6203312018001/ddc307e2-0354-47b3-861c-073c2c03cae5/dc628e2a-f1ab-49ff-8f65-700740a06609/28be398c-fdbe-40c4-904a-2e6f0e063ff5/default_video4000_5_1920x1080/media.m3u8?bc_token=NjRiODVmZDVfMTUyNTgxN2RjMTY0Y2UzMjJkNzZmZTA5ZTRmODgzMTgxYTE2OWYwOGI2OTU3YjE1MzQ1YjExOTk0NzJjZjU3YQ%3D%3D"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 123
[download] Destination: Ant-Man VS Atom (Marvel VS DC) [50934].fhls-4606.mp4
[debug] File locking is not supported. Proceeding without locking
[download] 100% of  554.69MiB in 00:03:05 at 2.99MiB/s
[debug] Invoking hlsnative downloader on "https://ssaimanifest.prod.boltdns.net/us-east-1/playback/once/v1/hls/v4/clear/6203312018001/ddc307e2-0354-47b3-861c-073c2c03cae5/dc628e2a-f1ab-49ff-8f65-700740a06609/28be398c-fdbe-40c4-904a-2e6f0e063ff5/4b45bfd7-2729-4aea-b477-f60642c9fd73/media.m3u8?bc_token=NjRiODVmZDVfMTBiNTUzNGNmMTcwNGM3NDM3NDhiOWQzNWJlYTFkYzgyMThkMDI3OGIyMmUxNmQ3YjdkM2IzMDc4NTY0YTJjOA%3D%3D"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 124
[download] Destination: Ant-Man VS Atom (Marvel VS DC) [50934].fhls-audio-3-en__Main_.mp4
[download] 100% of   31.73MiB in 00:02:14 at 242.04KiB/s
[debug] ffmpeg command line: ffprobe -show_streams "file:Ant-Man VS Atom (Marvel VS DC) [50934].fhls-audio-3-en__Main_.mp4"
[Merger] Merging formats into "Ant-Man VS Atom (Marvel VS DC) [50934].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:Ant-Man VS Atom (Marvel VS DC) [50934].fhls-4606.mp4" -i "file:Ant-Man VS Atom (Marvel VS DC) [50934].fhls-audio-3-en__Main_.mp4" -c copy -map "0:v:0" -map "1:a:0" "-bsf:a:0" aac_adtstoasc -movflags "+faststart" "file:Ant-Man VS Atom (Marvel VS DC) [50934].temp.mp4"
Deleting original file Ant-Man VS Atom (Marvel VS DC) [50934].fhls-audio-3-en__Main_.mp4 (pass -k to keep)
Deleting original file Ant-Man VS Atom (Marvel VS DC) [50934].fhls-4606.mp4 (pass -k to keep)
```

### PR DESCRIPTION
**Pull request by jkmartindale**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This PR makes the Rooster Teeth extractor grab ad-free streams instead of streams with embedded ads.

Based on the videos I've looked at, by default `video_data['attributes']['url']` contains ads and `video_data['links']['download']` does not. Preferring the latter over the former is enough to download ad-free streams.

Additionally, passing in a `Client-Type: web` header forces both URLs to be the same ad-free stream. This PR includes that header as well for some slight future-proofing.

The RoosterTeeth tests were failing due to incomplete attribute data in test cases so this PR fixes that too.

Fixes #7647 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by memmam** [comment]
Thank you so much for this, I'm trying to back up the entirety of RoosterTeeth in wake of yesterday's news and your branch has solved some significant issues I was having with 403/404 errors.

File: yt_dlp/extractor/roosterteeth.py [hunk 1]
```diff
@@ -133,10 +151,9 @@ def _real_extract(self, url):
 
         try:
             video_data = self._download_json(
-                api_episode_url + '/videos', display_id,
-                'Downloading video JSON metadata')['data'][0]
-            m3u8_url = video_data['attributes']['url']
-            # XXX: additional URL at video_data['links']['download']
+                api_episode_url + '/videos', display_id, 'Downloading video JSON metadata',
+                headers={'Client-Type': 'web'})['data'][0]
+            m3u8_url = traverse_obj(video_data, ('links', 'download')) or traverse_obj(video_data, ('attributes', 'url'))
```
Comment (bashonly) [inline yt_dlp/extractor/roosterteeth.py:156]: and move this down below the try/except block
```suggestion
            m3u8_url = traverse_obj(video_data, (
                (('links', 'download'), ('attributes', 'url')), {url_or_none}), get_all=False)
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/roosterteeth.py (hunk 1 above)
Comment (memmam) [reply yt_dlp/extractor/roosterteeth.py:156]: 'and'? Is part of your comment with any other changes missing?

**Comment by spschoen** [comment]
> Hey, if you can do any extra work on this, I'm getting lots of 403 errors trying to download the `https://roosterteeth.com/series/achievement-hunter` playlist

I don't want to clog up this PR with this (it may be an unrelated issue), but I do just want to say you can get an m3u8 link from the Rooster Teeth API that you can pass to `yt-dlp`.  This gets you the video file but not subtitles or any metadata.

e.g., if you're trying to download `https://roosterteeth.com/watch/red-vs-blue-season-3-thanksgiving-day-psa`, you'd go to `https://svod-be.roosterteeth.com/api/v1/episodes/red-vs-blue-season-3-thanksgiving-day-psa/videos` then grab the m3u8 link from `["data"][0]["attributes"]["url"]`.

That should allow you to download videos while a "proper" solution is implemented and merged, so long as you're okay with not having metadata or subs.

File: yt_dlp/extractor/roosterteeth.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/roosterteeth.py:156]: I am suggesting a change to this line of code, and that the line itself should also be moved down below the try/except block

**Comment by memmam** [comment]
The lack of subtitles going about it that way is a dealbreaker, I'm trying to back up the entire site in preparation for the eventual delisting of content as rights are sold off, if not the total shutdown of the site as things wind down over the next six months. Please keep me updated as to a proper solution.

**Comment by bashonly** [comment]
@memmam subtitles and playlist 403 errors are out-of-scope for this PR. Please refrain from off-topic comments

**Comment by bashonly** [comment]
Actually, my bad, the playlist 403 errors are caused by the `video_data['links']['download']` m3u8 URLs.  If passing the client header yields ad-free streams for the original m3u8 URLs we should only use those

File: yt_dlp/extractor/roosterteeth.py (hunk 1 above)
Comment (bashonly) [inline yt_dlp/extractor/roosterteeth.py:156]: `video_data['links']['download']` URLs were giving 403 errors
```suggestion
                headers={'Client-Type': 'web'})['data'][0]
            m3u8_url = video_data['attributes']['url']
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Comment by memmam** [comment]
@bashonly Yeah, I downloaded and built the fork with your changes, this PR fixes *some* of the 403 errors but seemingly not *all* of them, I'll take the hit as to that being not entirely related. My concern is that in the time since this PR was made, WB announced they're shutting Rooster Teeth down, so any development getting the extractor fully working should be a high priority so the site can be archived. I'll open an issue once this PR is merged.

**Comment by memmam** [comment]
@bashonly Yeah, I downloaded and built the fork with your first change, this PR fixes some of the 403 errors but seemingly not all of them, I'll take the hit as to that being not entirely related.

My concern is that in the time since this PR was made, WB announced they're shutting Rooster Teeth down, so any development getting the extractor fully working should be a high priority so the site can be archived. I'll try your new change, and if it's still a problem, open an issue once this PR is merged.

**Review by bashonly — APPROVED** [review]

**Comment by memmam** [comment]
Wait, either I misunderstood or, the line was updated but you also wanted it moved, which it wasn't?

**Comment by bashonly** [comment]
@memmam the line hasn't been changed. I retracted my original suggestion after your discovery that the new URLs were bad

**Comment by bashonly** [comment]
Some examples:

ad-free m3u8 url fetched by `client-type` header alone (`video_data['links']['download']` is not needed):
https://roosterteeth.com/watch/death-battle-season-10-antman

403 error if we use `video_data['links']['download']`:
https://www.roosterteeth.com/watch/achievement-hunter-achievement-hunter-fatality-walkthrough-deathstroke-lex-luthor-captain-marvel-green-lantern-and-wonder-woman

**Comment by vidyava** [comment]
Thank you for putting this PR together. I don't know if this would warrant an addition here or a separate PR, but I've added this line to the extractor to grab the upload date so videos could be sorted by time:
```
--- a/yt_dlp/extractor/roosterteeth.py
+++ b/yt_dlp/extractor/roosterteeth.py
@@ -70,6 +70,7 @@ def _extract_video_info(self, data):
             'episode_id': str_or_none(data.get('uuid')),
             'channel_id': attributes.get('channel_id'),
             'duration': int_or_none(attributes.get('length')),
+            'upload_date': attributes.get('original_air_date').split("T")[0].replace("-",""),
             'thumbnails': thumbnails,
```

**Comment by jkmartindale** [comment]
It's also trivial to add bonus feature extraction to the series extractor. Does that need a separate issue or can it go straight to another PR?

**Comment by bashonly** [comment]
@jkmartindale could you open a separate PR please?

@vidyava separate PR, I'll open it

**Comment by vidyava** [comment]
Just to expand on the 403 errors since this seems to be where they're being discussed: even with the updated header and urls some links still fail. An example:
[https://roosterteeth.com/watch/lets-play-2013-126](https://roosterteeth.com/watch/lets-play-2013-126)
returns a 403 error from yt-dlp despite the video playing in a browser, and going to either m3u8 url [returned by the API](https://svod-be.roosterteeth.com/api/v1/episodes/lets-play-2013-126/videos) directly gets:
```
[
  {
    "error_code": "VIDEO_NOT_PLAYABLE",
    "message": "The policy key provided does not permit this account or video, or the requested resource is inactive."
  }
]
```
Attempting the same process with [https://roosterteeth.com/watch/lets-play-2013-127](https://roosterteeth.com/watch/lets-play-2013-127) successfully downloads the m3u8 file and the video.  

Loading both with the network debugger open shows the former loads the stream from [https://rtv2-video.roosterteeth.com/uploads/videos/4e599900-19cd-4cf0-87fd-af80f6bbcf02/index.m3u8](https://rtv2-video.roosterteeth.com/uploads/videos/4e599900-19cd-4cf0-87fd-af80f6bbcf02/index.m3u8) and similar internal CDNs rather than brightcove or boltdns. I'm over my head on how to fall back when the API links to a non-existent playlist file, but providing info in case it's clear to anyone else.

**Comment by memmam** [comment]
@vidyava I think this needs to be opened as a separate issue. Could you do that, since you seem to have figured the problem out?

**Comment by NBruderman** [comment]
> @vidyava to add to this, by using the brightcove fallback bash provided in https://github.com/yt-dlp/yt-dlp/pull/9355#issuecomment-1984841014, you can see that the actual m3u8 is the same as the one loaded via the website:
> `https://players.brightcove.net/6203312018001/default_default/index.html?videoId=6311613926112`
> loads `https://rtv2-video.roosterteeth.com/uploads/videos/4e599900-19cd-4cf0-87fd-af80f6bbcf02/index.m3u8`
> 
> the brightcove video id is available in the `/videos` response under `attributes['uid']`
> 
> idk if these urls are known to be ad free in this case though. the empty `ad_breaks` array in the `/videos` could maybe hint at that

The thing is that this thread is already completed, might be worth opening a new one, so it won't get lost in the shuffle.

**Comment by bashonly** [comment]
@vidyava @NBruderman @Bl4Cc4t #9403

**Comment by NBruderman** [comment]
> It's also trivial to add bonus feature extraction to the series extractor. Does that need a separate issue or can it go straight to another PR?

Can you open a PR for that?

**Comment by Bl4Cc4t** [comment]
@NBruderman opened a PR above for bonus feature support

**Pull request merged**